Sources/TUSKit/TUSClient.swift

```diff
@@ -0,0 +1,523 @@
+            let tusError = TUSClientError.couldNotLoadData(underlyingError: error)
+            delegate?.fileError(error: tusError, client: self)
+            return []
+        }
+    }
+    
+    /// Schedule a single task if needed. Will decide what task to schedule for the metaData.
+    /// - Parameter metaData:The metaData the schedule.
+    private func scheduleTask(for metaData: UploadMetadata) throws {
+        guard let task = try taskFor(metaData: metaData, api: api, files: files, progressDelegate: self) else {
+            throw TUSClientError.uploadIsAlreadyFinished
+        }
+        uploads[metaData.id] = metaData
+        scheduler.addTask(task: task)
+    }
+    
+}
+
+extension TUSClient: SchedulerDelegate {
+    func didFinishTask(task: ScheduledTask, scheduler: Scheduler) {
+        switch task {
+        case let task as UploadDataTask:
+            handleFinishedUploadTask(task)
+        case let task as StatusTask:
+            handleFinishedStatusTask(task)
+        case let task as CreationTask:
+            handleCreationTask(task)
+        default:
+            break
+        }
+    }
+    
+    func handleCreationTask(_ creationTask: CreationTask) {
+        creationTask.metaData.errorCount = 0
+    }
+    
+    func handleFinishedStatusTask(_ statusTask: StatusTask) {
+        statusTask.metaData.errorCount = 0 // We reset errorcounts after a succesful action.
+        if statusTask.metaData.isFinished {
+            _ = try? files.removeFileAndMetadata(statusTask.metaData) // If removing the file fails here, then it will be attempted again at next startup.
+        }
+    }
+    
+    func handleFinishedUploadTask(_ uploadTask: UploadDataTask) {
+        uploadTask.metaData.errorCount = 0 // We reset errorcounts after a succesful action.
+        guard uploadTask.metaData.isFinished else { return }
+        
+        do {
+            try files.removeFileAndMetadata(uploadTask.metaData)
+        } catch let error {
+            let tusError = TUSClientError.couldNotDeleteFile(underlyingError: error)
+            delegate?.fileError(error: tusError, client: self)
+        }
+        
+        guard let url = uploadTask.metaData.remoteDestination else {
+            assertionFailure("Somehow uploaded task did not have a url")
+            return
+        }
+        
+        uploads[uploadTask.metaData.id] = nil
+        delegate?.didFinishUpload(id: uploadTask.metaData.id, url: url, context: uploadTask.metaData.context, client: self)
+    }
+    
+    func didStartTask(task: ScheduledTask, scheduler: Scheduler) {
+        guard let task = task as? UploadDataTask else { return }
+        
+        if task.metaData.uploadedRange == nil && task.metaData.errorCount == 0 {
+            delegate?.didStartUpload(id: task.metaData.id, context: task.metaData.context, client: self)
+        }
+    }
+    
+    func onError(error: Error, task: ScheduledTask, scheduler: Scheduler) {
+        let metaData: UploadMetadata
```

Comment:
Is it an idea to add `metaData: UploadMetaData` to `ScheduledTask`? It cleans up the following lines of code and removes the fatalError which we know will never be called anyway. Or is there a task that doesn't have metaData that I've missed?

Reply:
So ScheduledTask is part of Scheduler, which is its own system. It doesn't know anything about TUS on purpose, it's focusing only queueing async tasks. 

I could opt to add a payload to the `ScheduledTask` protocol, but then I'd either be downcasting anyway (like here), or we will enter the world of associatedtypes which has a heavy cost. I think a switch here is fine. Worst case we can abstract the switch to a task.

Reply:
I cleaned it up a bit with a little function and a guard statement. No more fatalerror. It will be there in a next commit